src/Core/Data/AbstractTypedCollection.php

```diff
@@ -0,0 +1,171 @@
+
+use Doctrine\Common\Collections\ArrayCollection;
+use PrestaShop\PrestaShop\Core\Exception\InvalidArgumentException;
+
+/**
+ * Class AbstractTypedCollection is an abstract collection class which checks
+ * that the inserted elements match the requested type.
+ */
+abstract class AbstractTypedCollection extends ArrayCollection
+{
+    /**
+     * Define the type of the elements contained in the collection.
+     * Example: for a ProductCollection you need to return Product::class
+     *
+     * @return string
+     */
+    abstract protected function getType();
+
+    /**
+     * AbstractTypedCollection constructor.
+     *
+     * @param array $elements
+     *
+     * @throws InvalidArgumentException
+     */
+    public function __construct(array $elements = array())
+    {
+        $this->checkElementsType($elements);
+        parent::__construct($elements);
+    }
+
+    /**
+     * @param mixed $element
+     *
+     * @return bool
+     *
+     * @throws InvalidArgumentException
+     */
+    public function removeElement($element)
+    {
+        $this->checkElementType($element);
+
+        return parent::removeElement($element);
+    }
+
+    /**
+     * @param mixed $offset
+     * @param mixed $value
+     *
+     * @return bool|void
+     *
+     * @throws InvalidArgumentException
+     */
+    public function offsetSet($offset, $value)
+    {
+        $this->checkElementType($value);
+
+        return parent::offsetSet($offset, $value);
+    }
+
+    /**
+     * @param mixed $element
+     *
+     * @return bool
+     *
+     * @throws InvalidArgumentException
+     */
+    public function contains($element)
+    {
+        $this->checkElementType($element);
+
+        return parent::contains($element);
+    }
+
+    /**
+     * @param mixed $element
+     *
+     * @return bool|false|int|string
+     *
+     * @throws InvalidArgumentException
+     */
+    public function indexOf($element)
+    {
+        $this->checkElementType($element);
+
+        return parent::indexOf($element);
+    }
+
+    /**
+     * @param mixed $key
+     * @param mixed $value
+     *
+     * @throws InvalidArgumentException
+     */
+    public function set($key, $value)
+    {
+        $this->checkElementType($value);
+
+        parent::set($key, $value);
+    }
+
+    /**
+     * @param mixed $element
+     *
+     * @return bool
+     *
+     * @throws InvalidArgumentException
+     */
+    public function add($element)
+    {
+        $this->checkElementType($element);
+
+        return parent::add($element);
+    }
+
+    /**
+     * @param array $elements
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function checkElementsType(array $elements)
+    {
+        foreach ($elements as $element) {
+            $this->checkElementType($element);
+        }
+    }
+
+    /**
+     * @param mixed $element
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function checkElementType($element)
```

Comment:
why not in private for the 2 last functions?

If people want to change this behavior, they'd better implement their own collection, isn't it?

Reply:
I hesitated, but ok let's make it private